src/monitor/node_active_protocol.c

```diff
@@ -1259,6 +1260,120 @@ perform_failover(PG_FUNCTION_ARGS)
 }
 
 
+/*
+ * promote promotes a given target node in a group.
+ */
+Datum
+perform_promotion(PG_FUNCTION_ARGS)
+{
+	text *formationIdText = PG_GETARG_TEXT_P(0);
+	char *formationId = text_to_cstring(formationIdText);
+
+	text *nodeNameText = PG_GETARG_TEXT_P(1);
+	char *nodeName = text_to_cstring(nodeNameText);
+
+	List *groupNodesList = NULL;
+	int totalNodesCount = 0;
+
+	AutoFailoverNode *currentNode = NULL;
+
+	checkPgAutoFailoverVersion();
+
+	currentNode = GetAutoFailoverNodeByName(formationId, nodeName);
+
+	if (currentNode == NULL)
+	{
+		ereport(ERROR,
+				(errmsg("node \"%s\" is not registered in formation \"%s\"",
+						nodeName, formationId)));
+	}
+
+	LockFormation(formationId, ShareLock);
+	LockNodeGroup(formationId, currentNode->groupId, ExclusiveLock);
+
+	/*
+	 * If the current node is the primary, that's done.
+	 */
+	if (IsCurrentState(currentNode, REPLICATION_STATE_SINGLE) ||
+		IsCurrentState(currentNode, REPLICATION_STATE_PRIMARY))
+	{
```

Comment:
minor: might be good to have  `ereport(NOTICE, .. node is already primary ..)` kind of information to the user

Reply:
Added.